Fix input/hotkey bugs

SuuperW opened the pull request:
Fix: Hotkey overrides input would cause an input to continue after a button was released if the input is bound to a combination such as Shift+A and an "internal hotkey" is bound to a single button (such as A). Also input bound to Alt would not have the alt ignored by menus.

Fix: Input overrides hotkeys did not work if both input and hotkey were bound to a combination such as Shift+A.

Fix: Controller inputs bound to a modifier key combination would not be triggered if additional modifier keys were also pressed.

Fix: autofire/autohold hotkeys would sometimes activate buttons that were already being held when the hotkey was pressed.

[//]: # "Apart from the mandatory license signature, these tasks are optional, but doing them could save reviewers some time and get the PR merged sooner."
Check if completed:
- [x] I have run any relevant test suites
- [x] I, the commit author, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-06-22) and am compliant

Morilli commented on src/BizHawk.Client.Common/Controller.cs:
```diff
@@ -59,11 +59,20 @@ public bool AnyInputHeld
+		/// This means Shift+A will return true if an input is bound to either A or Shift+A.
+		/// </summary>
+		public bool HasBinding(string button)
+		{
+			HashSet<string> buttons = button.Split('+').ToHashSet();
+			return _bindings
 				.SelectMany(kvp => kvp.Value)
-				.Any(boundButton => boundButton == button);
+				.Any((boundCombination) => {
+					HashSet<string> boundButtons = boundCombination.Split('+').ToHashSet();
+					return boundButtons.IsSubsetOf(buttons);
+				});
+		}
```
I don't think this logic is fully correct. When an input is bound to e.g. `Ctrl+E` and a hotkey is bound to `Ctrl+Shift+E` and "Input overrides Hotkeys" is active, pressing `Ctrl+Shift+E` will do nothing.

SuuperW replied:
In my testing, `Ctrl+Shift+E` will press the controller input and not trigger the hotkey, which is the expected behavior.
Inputs work even when pressing extra modifier keys because (from Discord)
> If we didn't consider Shift+A as also A then you could not press whatever is bound to Shift and then whatever is bound to A while still holding shift.

Morilli replied:
Have you tested this with `Input overrides Hotkeys`? Because I get nothing.

Morilli commented:
It's pretty hard to verify things are correct now because nothing defines "correct". Let me recap a couple things that happen.

If a key combination like `Ctrl+Shift+A` is pressed, it will now activate controller bindings for `Ctrl`, `Shift`, `A`, `Ctrl+Shift`, `Ctrl+A`, `Shift+A` and `Ctrl+Shift+A`. Only a hotkey bound to `Ctrl+Shift+A` will be triggered though.

"Internal" hotkeys are a separate group from normal hotkeys. If any key press activates a normal hotkey then no internal hotkey will be triggered, regardless of priority setting, except for the `Frame Inch` hotkey binding. The `Frame Inch` exception also means that binding a key combination to `Frame Inch` and some controller input will still have it go through in "Hotkeys override Input" mode.

The "Hotkeys override Input" branch is especially difficult for me to understand. Am I correct that the following code is functionally equivalent to what the code is currently doing?
```csharp
if (ie.EventType is InputEventType.Press)
{
	var handled = triggers.Aggregate(false, (current, trigger) => current | CheckHotkey(trigger));
	if (!handled)
	{
		hotkeyCoalescer.Receive(ie);
		if (!triggers.Exists(IsInternalHotkey))
		{
			finalHostController.Receive(ie);
		}
	}
}
else
{
	hotkeyCoalescer.Receive(ie);
	finalHostController.Receive(ie);
}
```

SuuperW commented:
> It's pretty hard to verify things are correct now because nothing defines "correct".

Yeah, we really should have some tests for this.

> If a key combination like `Ctrl+Shift+A` is pressed, it will now activate controller bindings for `Ctrl`, `Shift`, `A`, `Ctrl+Shift`, `Ctrl+A`, `Shift+A` and `Ctrl+Shift+A`. Only a hotkey bound to `Ctrl+Shift+A` will be triggered though.

This is correct, with the caveat that you cannot press `Ctrl+Shift+A` without first pressing `Ctrl+Shift` and before that either `Ctrl` or `Shift`, so hotkeys bound to those may trigger in the process. `A` must be pressed last or `Ctrl+Shift+A` won't happen for a hotkey, but a controller binding would be activated when pressing the keys in any order.
 
> "Internal" hotkeys are a separate group from normal hotkeys. If any key press activates a normal hotkey then no internal hotkey will be triggered, regardless of priority setting, except for the `Frame Inch` hotkey binding. The `Frame Inch` exception also means that binding a key combination to `Frame Inch` and some controller input will still have it go through in "Hotkeys override Input" mode.

Interesting. I had not looked that much into the internal hotkeys.
I think that system should be re-worked. One reason for this is that if you set a very low emulation speed, the frame advance hotkey is unreliable. So is pausing. (because you're likely to press and release between loops, so the loop will never see it pressed)

> The "Hotkeys override Input" branch is especially difficult for me to understand. Am I correct that the following code is functionally equivalent to what the code is currently doing?

Yeah lots of this code is tough to read. I believe your code is equivalent.

YoshiRulz commented:
> [...] if you set a very low emulation speed, the frame advance hotkey is unreliable. So is pausing. (because you're likely to press and release between loops, so the loop will never see it pressed)

#2959?

Morilli commented:
> if you set a very low emulation speed, the frame advance hotkey is unreliable. So is pausing.

I can't reproduce that with the pausing hotkey. However for frame advance you can use frame inch instead as that immediately pauses instead of waiting for the actual frame loop.

SuuperW commented:
Oh pause isn't an internal hotkey. Huh.
I must have been using the frame advance button to try to pause when I experienced that.

SuuperW commented:
Tests and additional fixes.

> "Internal" hotkeys are a separate group from normal hotkeys. If any key press activates a normal hotkey then no internal hotkey will be triggered

This is also fixed.

Rebased onto my undo_redo branch (has its own pending PR) in order to make use of test code there.

YoshiRulz commented on src/BizHawk.Client.Common/input/IInputSource.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace BizHawk.Client.Common
+{
+	public interface IInputSource
```
Name could be more specific

SuuperW commented on src/BizHawk.Tests.Client.Common/InputManagerTests.cs:
```diff
@@ -0,0 +1,500 @@
+		{
+			Context c = MakeContext();
+			InputManager manager = c.manager;
+			FakeInputSource source = new FakeInputSource();
```
Disagree. `[TestInitialize]` is never better, as it is more difficult to understand, especially if one doesn't know to look for that method. Imo, all the code necessary to understand a test should be visible when looking at the test method. ("visible" includes any methods you can see being used; you see it is used so you can navigate to it if needed)

SuuperW commented on src/BizHawk.Tests.Client.Common/InputManagerTests.cs:
```diff
@@ -0,0 +1,500 @@
+
+		private static readonly int PRIORITY_BOTH = 0;
+		private static readonly int PRIORITY_INPUT = 1;
+		private static readonly int PRIORITY_HOTKEY = 2;
```
Idk either, but it didn't seem important to add it here. I'll add a commit for that since you suggested it though.

SuuperW commented:
I realized that I messed up mouse coordinate transforming here. I've added a new commit 29ed1b3 that fixes that. (It's weird that we translate screen coordinates to client coordinates then to -10k to +10k range and back to client coordinates again but I assume there's a reason for that somewhere.)
I'll give you a chance to review the updates before I merge, @YoshiRulz.

YoshiRulz commented on src/BizHawk.Client.Common/inputAdapters/InputManager.cs:
```diff
@@ -147,5 +137,113 @@ private static AutofireController BindToDefinitionAF(
+		/// <param name="processUnboundInput">Events that did not do anything are forwarded out here.
+		/// This allows things like Windows' standard alt hotkeys (for menu items) to be handled by the
+		/// caller if the input didn't alrady do something else.</param>
+		public void ProcessInput(IInputSource source, Func<string, bool> processHotkey, Config config, Action<InputEvent> processUnboundInput)
```
Not RMB. RMouse X/Y is the relative mouse mechanism introduced for DOOM/DOSBox. I just checked at https://github.com/TASEmulators/BizHawk/commit/38a69e89e7bfdb418a5304cbad3ed49defd36f17 and it seems to work the same.
